File: cpp_build/source/calculus/calculus.cpp
#include "calculus/calculus.h"

#include <system/multi_array.h>
#include <system/math.h>
#include <system/exceptions.h>
#include <system/details/lambda_capture_holder.h>
#include <system/collections/list.h>
#include <functional>

#include "computing/basic.h"

namespace Sigmath {

RTTI_INFO_IMPL_HASH(339793155u, ::Sigmath::Calculus, ThisTypeBaseTypesInfo);

double Calculus::Derivative(System::Func<double, double> f, double x)
{
    const double dx = 1e-9;
    // small increment to find the derivative
    double y1 = f(x);
    double y2 = f(x + dx);
    double derivative = (y2 - y1) / dx;
    return derivative;
}

double Calculus::Integral(System::Func<double, double> f, double a, double b, int32_t n)
{
    double h = (b - a) / n;
    double integral = 0;
    for (int32_t i = 1; i < n; i++)
    {
        double x = a + i * h;
        integral += f(x);
    }
    integral += (f(a) + f(b)) / 2;
    integral *= h;
    return integral;
}

double Calculus::Limit(System::Func<double, double> f, System::Func<double, double> g, double x)
{
    const int32_t maxIterations = 1000;
    const double tolerance = 1e-9;
    double fx = f(x);
    double gx = g(x);
    double limit = fx / gx;
    for (int32_t i = 0; i < maxIterations; i++)
    {
        double fx1 = Derivative(f, x);
        double gx1 = Derivative(g, x);
        double newLimit = fx1 / gx1;
        if (Basic::Abs(newLimit - limit) < tolerance)
        {
            return newLimit;
        }
        limit = newLimit;
        x += 1e-9;
    }
    throw System::Exception(u"Limit does not exist or cannot be computed with L'Hopital's Rule.");
}

double Calculus::FindRoot(System::Func<double, double> f, double a, double b, double epsilon /* = 1e-6*/)
{
    if (f(a) * f(b) > 0)
    {
        throw System::ArgumentException(u"Function must have opposite signs at interval endpoints");
    }
    
    double c = (a + b) / 2;
    
    while (System::Math::Abs(f(c)) > epsilon)
    {
        if (f(a) * f(c) < 0)
        {
            b = c;
        }
        else
        {
            a = c;
        }
        
        c = (a + b) / 2;
    }
    
    return c;
}

int32_t Calculus::Factorial(int32_t n)
{
    int32_t result = 1;
    
    for (int32_t i = 2; i <= n; i++)
    {
        result *= i;
    }
    
    return result;
}

System::Func<System::ArrayPtr<double>, double> Calculus::PartialDerivative(System::Func<System::ArrayPtr<double>, double> f, int32_t index, double h /* = 1e-6*/)
{
    System::Details::LambdaCaptureHolder<System::Func<System::ArrayPtr<double>, double>> _lch_f = f;
    System::Func<System::ArrayPtr<double>, double>& _f = _lch_f.GetCapture();
    System::Details::LambdaCaptureHolder<int32_t> _lch_index = index;
    int32_t& _index = _lch_index.GetCapture();
    System::Details::LambdaCaptureHolder<double> _lch_h = h;
    double& _h = _lch_h.GetCapture();
    return static_cast<System::Func<System::ArrayPtr<double>, double>>(static_cast<std::function<double(System::ArrayPtr<double> x)>>([_lch_index, &_index, _lch_h, &_h, _lch_f, &_f](System::ArrayPtr<double> x) -> double
    {
        auto xPlus = System::DynamicCast<System::Array<double>>(x->Clone());
        xPlus[_index] += _h;
        
        auto xMinus = System::DynamicCast<System::Array<double>>(x->Clone());
        xMinus[_index] -= _h;
        
        return (_f(xPlus) - _f(xMinus)) / (2 * _h);
    })).template AddHeldVariable<System::Func<System::ArrayPtr<double>, double>>("_index", _index).template AddHeldVariable<System::Func<System::ArrayPtr<double>, double>>("_h", _h).template AddHeldVariable<System::Func<System::ArrayPtr<double>, double>>("_f", _f);
}

System::Func<System::ArrayPtr<double>, System::ArrayPtr<double>> Calculus::Gradient(System::Func<System::ArrayPtr<double>, double> f, double h /* = 1e-6*/)
{
    System::Details::LambdaCaptureHolder<System::Func<System::ArrayPtr<double>, double>> _lch_f = f;
    System::Func<System::ArrayPtr<double>, double>& _f = _lch_f.GetCapture();
    System::Details::LambdaCaptureHolder<double> _lch_h = h;
    double& _h = _lch_h.GetCapture();
    return static_cast<System::Func<System::ArrayPtr<double>, System::ArrayPtr<double>>>(static_cast<std::function<System::ArrayPtr<double>(System::ArrayPtr<double> x)>>([_lch_f, &_f, _lch_h, &_h](System::ArrayPtr<double> x) -> System::ArrayPtr<double>
    {
        auto gradient = System::MakeArray<double>(x->get_Length(), 0);
        
        for (int32_t i = 0; i < x->get_Length(); i++)
        {
            gradient[i] = PartialDerivative(_f, i, _h)(x);
        }
        
        return gradient;
    })).template AddHeldVariable<System::Func<System::ArrayPtr<double>, System::ArrayPtr<double>>>("_f", _f).template AddHeldVariable<System::Func<System::ArrayPtr<double>, System::ArrayPtr<double>>>("_h", _h);
}

System::Func<System::ArrayPtr<double>, System::ArrayPtr<std::vector<double>>> Calculus::Hessian(System::Func<System::ArrayPtr<double>, double> f, double h /* = 1e-6*/)
{
    System::Details::LambdaCaptureHolder<System::Func<System::ArrayPtr<double>, double>> _lch_f = f;
    System::Func<System::ArrayPtr<double>, double>& _f = _lch_f.GetCapture();
    System::Details::LambdaCaptureHolder<double> _lch_h = h;
    double& _h = _lch_h.GetCapture();
    return static_cast<System::Func<System::ArrayPtr<double>, System::ArrayPtr<std::vector<double>>>>(static_cast<std::function<System::ArrayPtr<std::vector<double>>(System::ArrayPtr<double> x)>>([_lch_f, &_f, _lch_h, &_h](System::ArrayPtr<double> x) -> System::ArrayPtr<std::vector<double>>
    {
        int32_t n = x->get_Length();
        auto hessian = System::MakeArray<std::vector<double>>(n);
        RESIZE_MARRAY_INIT(i1, hessian->data(), n, 0)
        RESIZE_MARRAY_END()
        
        
        for (int32_t i = 0; i < n; i++)
        {
            for (int32_t j = 0; j < n; j++)
            {
                double fxx = PartialDerivative(PartialDerivative(_f, i, _h), j, _h)(x);
                hessian[i][j] = fxx;
            }
        }
        
        return hessian;
    })).template AddHeldVariable<System::Func<System::ArrayPtr<double>, System::ArrayPtr<std::vector<double>>>>("_f", _f).template AddHeldVariable<System::Func<System::ArrayPtr<double>, System::ArrayPtr<std::vector<double>>>>("_h", _h);
}

System::ArrayPtr<double> Calculus::FindCriticalPoints(System::Func<double, double> f, double a, double b, double epsilon /* = 1e-6*/)
{
    auto criticalPoints = System::MakeObject<System::Collections::Generic::List<double>>();
    
    // Evaluate function at endpoints of interval
    double fa = f(a);
    double fb = f(b);
    
    // Check if endpoints are critical points
    if (System::Math::Abs(fa) < epsilon)
    {
        criticalPoints->Add(a);
    }
    if (System::Math::Abs(fb) < epsilon)
    {
        criticalPoints->Add(b);
    }
    
    // Find critical points in the interval
    double x = a;
    double fx = f(a);
    
    while (x < b)
    {
        double xNext = x + epsilon;
        double fxNext = f(xNext);
        
        // Check if the current point is a critical point
        if (System::Math::Abs(fx) < epsilon && x > a && x < b)
        {
            criticalPoints->Add(x);
        }
        
        // Check if the function changes sign between x and xNext
        if (fx * fxNext < 0)
        {
            // Use bisection method to find a root
            double xLeft = x;
            double xRight = xNext;
            
            while (xRight - xLeft > epsilon)
            {
                double xMid = (xLeft + xRight) / 2;
                double fxMid = f(xMid);
                
                if (System::Math::Abs(fxMid) < epsilon)
                {
                    criticalPoints->Add(xMid);
                    break;
                }
                
                if (fx * fxMid < 0)
                {
                    xRight = xMid;
                }
                else
                {
                    xLeft = xMid;
                }
            }
        }
        
        x = xNext;
        fx = fxNext;
    }
    
    return criticalPoints->ToArray();
}

double Calculus::RiemannSum(System::Func<double, double> f, double a, double b, int32_t n, Calculus::RiemannSumType type /* = Sigmath::Calculus::RiemannSumType::Left*/)
{
    double dx = (b - a) / n;
    double sum = 0;
    
    for (int32_t i = 0; i < n; i++)
    {
        double x = a + i * dx;
        double fx = f(x);
        
        switch (type)
        {
            case Sigmath::Calculus::RiemannSumType::Left:
                sum += fx * dx;
                break;
            
            case Sigmath::Calculus::RiemannSumType::Right:
                sum += f(x + dx) * dx;
                break;
            
            case Sigmath::Calculus::RiemannSumType::Midpoint:
                sum += f(x + dx / 2) * dx;
                break;
            
            case Sigmath::Calculus::RiemannSumType::Trapezoidal:
                sum += (fx + f(x + dx)) * dx / 2;
                break;
            
        }
    }
    
    return sum;
}

double Calculus::DefiniteIntegral(System::Func<double, double> f, double a, double b, int32_t n /* = 1000*/)
{
    return RiemannSum(f, a, b, n, Sigmath::Calculus::RiemannSumType::Midpoint);
}

} // namespace Sigmath

File: Sigmath/Sigmath/Sigmath.ConvertedToPython/calculus/calculus.py
from System import *
from System.Collections.Generic import *

class Calculus(object):
	def Derivative(f, x):
		dx = 1e-9 # small increment to find the derivative
		y1 = Calculus.f(x)
		y2 = Calculus.f(x + dx)
		derivative = (y2 - y1) / dx
		return derivative

	Derivative = staticmethod(Derivative)

	def Integral(f, a, b, n):
		h = (b - a) / n
		integral = 0
		i = 1
		while i < n:
			x = a + i * h
			integral += Calculus.f(x)
			i += 1
		integral += (Calculus.f(a) + Calculus.f(b)) / 2
		integral *= h
		return integral

	Integral = staticmethod(Integral)

	def Limit(f, g, x):
		maxIterations = 1000
		tolerance = 1e-9
		fx = Calculus.f(x)
		gx = Calculus.g(x)
		limit = fx / gx
		i = 0
		while i < maxIterations:
			fx1 = Calculus.Derivative(f, x)
			gx1 = Calculus.Derivative(g, x)
			newLimit = fx1 / gx1
			if Basic.Abs(newLimit - limit) < tolerance:
				return newLimit
			limit = newLimit
			x += 1e-9
			i += 1
		raise Exception("Limit does not exist or cannot be computed with L'Hopital's Rule.")

	Limit = staticmethod(Limit)

	def FindRoot(f, a, b, epsilon):
		if Calculus.f(a) * Calculus.f(b) > 0:
			raise ArgumentException("Function must have opposite signs at interval endpoints")
		c = (a + b) / 2
		while Math.Abs(Calculus.f(c)) > epsilon:
			if Calculus.f(a) * Calculus.f(c) < 0:
				b = c
			else:
				a = c
			c = (a + b) / 2
		return c

	FindRoot = staticmethod(FindRoot)

	def Factorial(n):
		result = 1
		i = 2
		while i <= n:
			result *= i
			i += 1
		return result

	Factorial = staticmethod(Factorial)

	def PartialDerivative(f, index, h):
		return 

	PartialDerivative = staticmethod(PartialDerivative)

	def Gradient(f, h):
		return 

	Gradient = staticmethod(Gradient)

	def Hessian(f, h):
		return 

	Hessian = staticmethod(Hessian)

	def FindCriticalPoints(f, a, b, epsilon):
		criticalPoints = List[Double]()
		# Evaluate function at endpoints of interval
		fa = Calculus.f(a)
		fb = Calculus.f(b)
		# Check if endpoints are critical points
		if Math.Abs(fa) < epsilon:
			criticalPoints.Add(a)
		if Math.Abs(fb) < epsilon:
			criticalPoints.Add(b)
		# Find critical points in the interval
		x = a
		fx = Calculus.f(a)
		while x < b:
			xNext = x + epsilon
			fxNext = Calculus.f(xNext)
			# Check if the current point is a critical point
			if Math.Abs(fx) < epsilon and x > a and x < b:
				criticalPoints.Add(x)
			# Check if the function changes sign between x and xNext
			if fx * fxNext < 0:
				# Use bisection method to find a root
				xLeft = x
				xRight = xNext
				while xRight - xLeft > epsilon:
					xMid = (xLeft + xRight) / 2
					fxMid = Calculus.f(xMid)
					if Math.Abs(fxMid) < epsilon:
						criticalPoints.Add(xMid)
						break
					if fx * fxMid < 0:
						xRight = xMid
					else:
						xLeft = xMid
			x = xNext
			fx = fxNext
		return criticalPoints.ToArray()

	FindCriticalPoints = staticmethod(FindCriticalPoints)

	class RiemannSumType(object):
		def __init__(self):

	def RiemannSum(f, a, b, n, type):
		dx = (b - a) / n
		sum = 0
		i = 0
		while i < n:
			x = a + i * dx
			fx = Calculus.f(x)
			if type == RiemannSumType.Left:
				sum += fx * dx
			elif type == RiemannSumType.Right:
				sum += Calculus.f(x + dx) * dx
			elif type == RiemannSumType.Midpoint:
				sum += Calculus.f(x + dx / 2) * dx
			elif type == RiemannSumType.Trapezoidal:
				sum += (fx + Calculus.f(x + dx)) * dx / 2
			i += 1
		return sum

	RiemannSum = staticmethod(RiemannSum)

	def DefiniteIntegral(f, a, b, n):
		return Calculus.RiemannSum(f, a, b, n, RiemannSumType.Midpoint)

	DefiniteIntegral = staticmethod(DefiniteIntegral)